File: array/34_Find_FirstLast_Element/34.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def getRightBorder(nums: List[int], target:int) -> int:
            left, right = 0, len(nums)
            rightBorder = -2
            while left < right:
                middle = left + (right - left) // 2
                if nums[middle] > target:
                    right = middle
                else:
                    left = middle + 1
                    rightBorder = left
            return rightBorder

        def getLeftBorder(nums: List[int], target:int) -> int:
            left, right = 0, len(nums)
            leftBorder = -2
            while left < right:
                middle = left + (right - left) // 2
                if nums[middle] < target:
                    left = middle + 1
                else:
                    right = middle
                    leftBorder = right
            return leftBorder

        leftBorder = getLeftBorder(nums, target)
        rightBorder = getRightBorder(nums, target)
        if leftBorder == -2 or rightBorder == -2:
            return [-1, -1]
        if rightBorder - leftBorder > 0:
            return [leftBorder, rightBorder - 1]
        return [-1, -1]

File: array/34_Find_FirstLast_Element/34.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int leftBorder = getLeftBorder(nums, target);
        int rightBorder = getRightBorder(nums, target);
        if (leftBorder == -2 || rightBorder == -2) {
            return {-1, -1};
        }
        if (rightBorder - leftBorder > 0) {
            return {leftBorder, rightBorder - 1};
        }
        return {-1, -1};
    }

private:
    int getRightBorder(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size();
        int rightBorder = -2;
        while (left < right) {
            int middle = left + ((right - left) >> 1);
            if (nums[middle] > target) {
                right = middle;
            } else {
                left = middle + 1;
                rightBorder = left;
            }
        }
        return rightBorder;
    }

    int getLeftBorder(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size();
        int leftBorder = -2;
        while (left < right) {
            int middle = left + ((right - left) >> 1);
            if (nums[middle] < target) {
                left = middle + 1;
            } else {
                right = middle;
                leftBorder = right;
            }
        }
        return leftBorder;
    }
};